src/bin/pg_autoctl/keeper_listener.c

```diff
@@ -0,0 +1,398 @@
+/*
+ * src/bin/pg_autoctl/keeper_listener.c
+ *	 The keeper listener is a process that reads commands from a PIPE and then
+ *	 synchronoulsy writes to the same PIPE the result of running given
+ *	 commands. This process is used to implement FSM transitions when running
+ *	 in --disable-monitor mode.
+ *
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the PostgreSQL License.
+ *
+ *   One reason to do use a separate process with a PIPE to enable two-way sync
+ *   communication is that we don't want the postgres processes to inherit from
+ *   the HTTPd server socket and other pg_autoctl context; so the clean way is
+ *   to have a process hierarchy where the HTTPd service is not the parent of
+ *   the Postgres related activity.
+ *
+ *   pg_autoctl run
+ *    - keeper run loop   [monitor enabled]
+ *    - httpd server      [all cases]
+ *    - listener          [all cases] [published API varies]
+ *      - pg_autoctl do fsm assign single
+ *      - pg_autoctl do fsm assign wait_primary
+ *      - pg_autoctl enable maintenance
+ *      - pg_autoctl disable maintenance
+ *    - postgres -p 5432 -h localhost -k /tmp
+ *
+ *   We still want the `postgres` process to run as a chile of the main
+ *   pg_autoctl service. When PostgreSQL is started by a listener command, this
+ *   means we also need a communication/execution channel with the pg_autoctl
+ *   parent process.
```

Comment:
This part has not been implemented, it's unclear whether it's necessary to add that level of complexity. It's also something that could be done separately later in another PR.